core/blockchain.go

```diff
@@ -1462,10 +1462,15 @@ func (bc *BlockChain) insertChain(chain types.Blocks, verifySeals, catalystMode
 		// Falls through to the block import
 	}
 	switch {
-	// First block is pruned, insert as sidechain and reorg only if TD grows enough
-	case errors.Is(err, consensus.ErrPrunedAncestor):
+	// We're in catalyst mode and the parent is pruned, try to recover the parent state
+	case catalystMode && errors.Is(err, consensus.ErrPrunedAncestor):
+		log.Debug("Pruned ancestor", "number", block.Number(), "hash", block.Hash())
+		return it.index, bc.recoverAncestors(block)
+
+		// First block is pruned, insert as sidechain and reorg only if TD grows enough
+	case !catalystMode && errors.Is(err, consensus.ErrPrunedAncestor):
 		log.Debug("Pruned ancestor, inserting as sidechain", "number", block.Number(), "hash", block.Hash())
-		return bc.insertSideChain(block, it)
+		return bc.insertSideChain(block, it, catalystMode)
```

Comment:
`insertSideChain` only works for non-catalyst mode, so that we can always pass `False`? (nitpick)